### DISCUSSION TITLE
How can I use `pytest` within an embedded Python environment with `__debug__ == False`?

### DISCUSSION THREAD

**Discussion by bradleyharden** [pytest-dev/pytest#12460] [Q&A] [answered]
I'm trying to use `pytest` within an application that uses Python as an embedded scripting language. However, it looks like the application invokes Python with the equivalent of `-O`, and `__debug__` is `False`. Thus, none of my asserts actually raise an `AssertionError`, and much of `pytest` is rendered useless.

What options do I have here? I suppose I could manually raise `AssertionError`s, but that gives up a lot of the benefits of assertion rewriting. It would at least give me a proper unit test framework, but it's certainly less than ideal. Are there other options?

**Comment by bradleyharden** [comment]
As I've been looking into this, I'm confused why `pytest` doesn't work here. According to [this](http://pybites.blogspot.com/2011/07/behind-scenes-of-pytests-new-assertion.html) blog post and a quick skim of the code, it seems like assertion rewriting should automatically convert all existing `assert` statements to a `raise AssertionError` equivalent. That's pretty much exactly what I need. Yet it doesn't work? Why?

**Reply by RonnyPfannschmidt** [reply]
for sanity, you might want to have your test print its own `dis.dis`

**Reply by The-Compiler** [reply]
I'm guessing whatever pytest does happens *after* Python already did it's optimization. In Fact, pytest warns about this:

> /usr/lib/python3.12/site-packages/_pytest/config/__init__.py:1286: PytestConfigWarning: assertions not in test modules or plugins will be ignored because assert statements are not executed by the underlying Python interpreter (are you using python -O?)

**Reply by bradleyharden** [reply]
Is that warning relevant here? Specifically

> assertions **not** in test modules or plugins will be ignored

Doesn't that suggest that any assertions that **are** in test modules will work correctly?

I'll look into `dis`.

**Reply by The-Compiler** [reply]
Ah, fair point. And indeed, with the assertion rewriting turned on, an assertion in a test module still fails. With `-O --assert=plain`, the warning changes to:

> /usr/lib/python3.12/site-packages/_pytest/config/__init__.py:1286: PytestConfigWarning: ASSERTIONS ARE NOT EXECUTED and FAILING TESTS WILL PASS.  Are you using python -O?

**Reply by bradleyharden** [reply]
@RonnyPfannschmidt and @The-Compiler, thanks for your help so far.

I left for vacation shortly after my last message, and I'm just now returning to this. I will dig in with `dis` and see if I can trace things through the `pytest` internals, but first I wanted to ask about one thing. I haven't mentioned this until now, because I didn't think it was relevant, but I'm starting to reevaluate that.

As I mentioned, I'm trying to run `pytest` within graphical application that uses Python as an embedded scripting language. The source of that application is not available to me, but I have gleaned some information by inspecting the Python environment I'm given. From that, I can tell that the Python console within the application is a custom-built, `PyQt4` object. Moreover, the author replaced `sys.std*` with custom objects that accept input from and print output to that console.

Those custom `sys.std*` objects actually caused me some trouble at first. They are not derived from [IOBase](https://docs.python.org/3/library/io.html#io.IOBase), so they lack several methods used by `pytest`, like `.isatty()` and `.fileno()`. I was able to monkey-patch the objects with:

```python
sys.stderr.fileno = lambda: 0
sys.stdout.isatty = lambda: False
sys.stdout.readable = lambda: False
sys.stdout.writable = lambda: True
```

But that only partially fixes the problem. As far as I can tell, the `_pytest.faulthandler.pytest_configure` method runs _unconditionally_ at `pytest` startup.

https://github.com/pytest-dev/pytest/blob/ac4189875562a6ce5b502c3dd87156c0367cf8f1/src/_pytest/faulthandler.py#L26-L38

It tries to call `os.dup` with a made up file descriptor, which unsurprisingly returns an error. As a ~major hack~ workaround, I also monkey-patched this function to simply call `faulthandler.enable()`. But that means the `stash` is missing `fault_handler_stderr_fd_key`. I don't know the internals of `pytest`, so I have no idea what the consequences of that might be.

After adding that monkey-patch, `pytest` seems to work. It correctly prints to the graphical console within the application, and it runs the tests. But perhaps this is causing assertion rewriting to silently fail? Could anyone with knowledge of the `pytest` internals comment?

**Reply by bradleyharden** [reply]
It looks like

```python
def test_failure():
    assert False

def test_dis():
    import dis
    print(dis.dis(test_failure))
```

gives

```
 61           0 LOAD_CONST               0 (None)
              2 RETURN_VALUE
None
```

**Reply by RonnyPfannschmidt** [reply]
That looks like assertion is disabled,

**Reply by bradleyharden** [reply]
~~After enabling `--debug` and looking through the logs, I see lines with~~

```
early skip of rewriting module
```

~~for my test modules. Do you have any idea why that would be the case? Maybe I'm using a strange or incorrect configuration?~~

Edit: Sorry, this is wrong. I misread the logs. Other modules are skipped, but not my test module.

**Reply by bradleyharden** [reply]
After some digging, it seems like my tests are not getting rewritten because of a strange type mismatch in `AssertionRewritingHook.find_spec`.

https://github.com/pytest-dev/pytest/blob/ac4189875562a6ce5b502c3dd87156c0367cf8f1/src/_pytest/assertion/rewrite.py#L103-L115

Specifically, the class of `spec.loader` does not seem to match `importlib.machinery.SourceFileLoader`. The following code

```python
print(f"{spec.loader.__class__}")
print(f"{importlib.machinery.SourceFileLoader}")
print(isinstance(spec.loader, importlib.machinery.SourceFileLoader))
```

yields

```
<class '_frozen_importlib_external.SourceFileLoader'>
<class '_frozen_importlib_external.SourceFileLoader'>
False
```

If I remove the `isinstance` check, my assertions are rewritten and everything works. Do you have any idea what might be going on?

**Reply by bradleyharden** [reply]
Interestingly,

```python
print(isinstance(spec.loader, importlib.abc.FileLoader))
print(isinstance(spec.loader, importlib.abc.SourceLoader))
```

yields

```
False
False
```

This really doesn't make any sense to me. Are there somehow two separate instances of `importlib` coexisting?

**Reply by bradleyharden** [reply]
After more digging, I can see some limited evidence that the application has modified parts of `importlib`. My current theory is that the various `find_spec` functions are returning an instance of a `SourceFileLoader` class that has since been replaced within `importlib` and is no longer reachable by import.

At this point, I'm pretty confident there's nothing any of you can do to help me. For the moment, my workaround is to modify `isinstance(spec.loader, importlib.machinery.SourceFileLoader` to be `spec.loader.__class__.__name__ == "SourceFileLoader"`. If anyone has a better idea, I would love to hear it.

**Reply by bradleyharden** [reply, marked as answer]
It looks like `sys.path_hooks[1]` returns instances of an old `FileFinder` class that is not reachable at `importlib.machinery.FileFinder`. The spec objects returned by its `find_spec` method also refer to an old version of `SourceFileLoader`. That's what causes the `isinstance` check to fail.

The simplest solution I've found is to extract a reference to the old class and replace the new class, i.e.
```python
importlib.machinery.SourceFileLoader = sys.path_hooks[1].__globals__["SourceFileLoader"]
```

As best I can tell, this doesn't cause any side effects.

This has been a strange adventure, but I think it ends here.

**Reply by RonnyPfannschmidt** [reply]
Thanks for providing the follow up 

For future sanity it May be helpful to work with your vendor so they can provide a better pytest integration out of the box

**Reply by bradleyharden** [reply]
Yeah, that would have been a lot easier, but for a number of reasons it probably wouldn't have happened. I'm trying to replace our existing system for tests, which essentially amounts to a very primitive unit testing framework. I can't really get buy-in for that without showing people how it could be better. And buy-in would be a prerequisite to requesting vendor support.

**Reply by RonnyPfannschmidt** [reply]
Having the vendor in the Loop might be a help 

Selling a duct-tape pytest integrate to a higher up May be harder than selling a update from the vendor 

But that might be undue optimism on my side
